js/packages/teams-ai/src/models/OpenAIModel.ts

```diff
@@ -282,15 +283,60 @@ export class OpenAIModel implements PromptCompletionModel {
     ): Promise<PromptResponse<string>> {
         const startTime = Date.now();
         const max_input_tokens = template.config.completion.max_input_tokens;
+
+        const chatCompletionTools: OpenAIFunction[] = [];
+        const chatCompletionToolHandlers: Map<string, ActionEntry<TurnState>> = memory.getValue('temp.toolHandlers');
+        const isToolsEnabled = template.config.completion.include_tools ?? false;
+
         const model =
             template.config.completion.model ??
             (this._useAzure
                 ? (this.options as AzureOpenAIModelOptions).azureDefaultDeployment
                 : (this.options as OpenAIModelOptions).defaultModel);
-        
-        // Check for legacy completion type  
+
+        // Check for legacy completion type
         if (template.config.type == 'completion') {
-            throw new Error(`The completion type 'completion' is no longer supported. Only 'chat' based models are supported.`);
+            throw new Error(
+                `The completion type 'completion' is no longer supported. Only 'chat' based models are supported.`
+            );
+        }
+
+        // If tools are enabled, massage data with action handlers to match schema
+        if (isToolsEnabled) {
+            if (!template.actions || template.actions.length == 0) {
+                return this.returnToolsError('noActions', undefined);
+            }
+
+            if (template.actions.length !== chatCompletionToolHandlers.size) {
+                return this.returnToolsError('lengthMismatch', undefined);
+            }
+
+            for (const action of template.actions!) {
+                if (chatCompletionToolHandlers.has(action.name)) {
+                    const toolsHandler = chatCompletionToolHandlers.get(action.name)!.handler;
+
+                    chatCompletionTools.push({
+                        name: action.name,
+                        handler: toolsHandler,
+                        description: action.description,
+                        parameters: action.parameters
+                    });
+                }
+            }
+
+            const formattedChatCompletionTools: ChatCompletionTool[] = [];
```

Comment:
for JS, you might not have to create this DS- we had to do it for python bc of the inflexibility with json objects

Reply:
Resolved 1:1 -- keeping for explicitness' sake :)